### DISCUSSION TITLE
Smartspacer Generic Weather not available in "Send Gadgetbridge data" menu

### DISCUSSION THREAD

**Discussion by pacjo** [breezy-weather/breezy-weather#949] [Q&A] [answered]
### Steps to reproduce

1. Update to 5.2.0
2. Try to add apps other then Gadgetbridge in "Send Gadgetbridge data" menu
3. See (well not an error) but incomplete list

### Expected behavior

1. All apps should appear

### Actual behavior

An incompleate list appears

### Weather source used

/

### Breezy Weather version

5.2.0

### Android version

Android 14

### Device

Android Studio Emulator / Redmi Note 10 Pro

### Other details

![Screenshot_20240415_200934](https://github.com/breezy-weather/breezy-weather/assets/56438628/a1700b5e-5f25-43e7-a6aa-e003d98cbb15)

This is probably a consequence of 015c1deccebfac8939dc2ef6e01c1f18ba5be6b9. 

I don't know how you're checking for available apps, I can add modify mine ([GenericWeather](https://github.com/pacjo/SmartspacerPlugins/tree/main/app/src/genericweather)) if needed.

There might be an issue with detecting the "support". I'm not registering any intent receivers. For me it's all handled by [Smartspacer](https://github.com/KieronQuinn/Smartspacer), with only code being [this](https://github.com/pacjo/SmartspacerPlugins/blob/f7dfa42678306688ec84d7f09235c3edad1ad660/app/src/genericweather/src/java/providers/GenericWeatherProvider.kt#L41).

Maybe a solution might be to register a dummy intent filter to make it visible to Breezy?

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open or closed issue.
- [X] I have written a short but informative title.
- [X] I read the [Frequently Asked Questions](https://github.com/breezy-weather/breezy-weather/blob/main/HELP.md) document and followed instructions if relevant.
- [X] I have updated the app to **[version 5.2.0 or later](https://github.com/breezy-weather/breezy-weather/releases/latest)**.
- [X] I wrote my request in English, and any screenshot I attached are also in English (language can be changed from Settings > Appearance), unless my issue only happens with a specific language
- [X] I understand that this app is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY, so I may not get a positive reply to my request, if any at all.

**Comment by papjul** [comment]
I didn't know you could get an intent without declaring it in the manifest.

You will need to declare it. See line 447 of how Gadgetbridge does it:
https://codeberg.org/Freeyourgadget/Gadgetbridge/src/branch/master/app/src/main/AndroidManifest.xml

**Comment by pacjo** [comment]
Thanks for a quick response. I changed this:
```
        <provider
            android:name="providers.GenericWeatherProvider"
            android:authorities="${applicationId}.providers.weather"
            android:exported="true"
            android:permission="com.kieronquinn.app.smartspacer.permission.ACCESS_SMARTSPACER_BROADCASTS" />
```
to this
```
<provider
            android:name="providers.GenericWeatherProvider"
            android:authorities="${applicationId}.providers.weather"
            android:exported="true"
            android:permission="com.kieronquinn.app.smartspacer.permission.ACCESS_SMARTSPACER_BROADCASTS" >
            <intent-filter>
                <action android:name="nodomain.freeyourgadget.gadgetbridge.ACTION_GENERIC_WEATHER" />
            </intent-filter>
        </provider>
```
and now it shows up in the settings, but I still don't get intents.

It's probably because Smartspacer app gets them normally, then just calling a function in my plugin with the intent. Can we do something about it? (...just maybe (if possible) an "other apps" button should be made available?)

**Comment by Releece** [comment]
I have the same issue.
Device: OnePlus 11
OS: Android 14 (OxygenOS 14.0.0.601(EX01))
Breezy Weather: 5.2.0
Smartspacer: 1.7.2

Before, in order to use Breezy Weather as a weather source for Smartspacer I had to install GadgetBridge (which I don't have any use for) just so I could enable the toggle to send data to GadgetBridge, and in turn Smartpacer would receive the data.

This was then updated to a more general toggle to share data with other apps. Solving the problem of needing to install Gadgetbridge and I could succesfully share data with Smartspacer after uninstalling GadgetBridge.

Now with the most recent update to be able to select which apps you want to share to, if I press 'Send GadgetBridge data' I get 'No compatible packages found', while Smartspacer + the Smartspacer plugin to fetch weather data from other sources are installed, as they have always been throughout these updates on my device.


Please fix kindly :) Thanks for all the work on Breezy

**Comment by papjul** [comment]
> now it shows up in the settings

> just maybe (if possible) an "other apps" button should be made available?

So you don't need that if it's already listed?

> but I still don't get intents

We changed two things in the way we send intents:
- We added the `WeatherSecondaryJson` extra, a new feature from Gadgetbridge
- We set the package to which we send the intent

Both seem unrelated to your issue to me, as Gadgetbridge developer confirmed this new method works. If you can see your app listed, the intent will be sent to your app.

> It's probably because Smartspacer app gets them normally, then just calling a function in my plugin with the intent

Which package is listed? Smartspacer or your plugin app? The manifest with the intent is in your app, not in Smartspacer app, right?

**Comment by pacjo** [comment]
Right, I'll try to explain again.

As per [Smartspacer docs](https://github.com/KieronQuinn/Smartspacer/wiki/8.-SDK:-Broadcast-Providers):

> Smartspacer handles registering the broadcast receiver for you, and will send your plugin's Broadcast Provider the raw intents so you can load any data from them, and act as required.

So intent should probably be sent to Smartspacer, and not GenericWeather. 

I see two solutions here:
- either allow any app to be selected in the menu (even if it doesn't register the intent filter) (this is the one I prefer)
- or intent filter could be added to the Smartspacer itself, but that would mean that multiple apps have to cooperate to restore functionality from the last version

> We changed two things in the way we send intents:

> We added the WeatherSecondaryJson extra, a new feature from Gadgetbridge
> We set the package to which we send the intent

WeatherSecondaryJson isn't the issue. I'll fix it in my app.

> Which package is listed? Smartspacer or your plugin app? The manifest with the intent is in your app, not in Smartspacer app, right?

In the intent filter? Mine. That's what I have control over and I'd prefer not to modify other projects just to make mine function.

Problem here is specific to how Smartspacer works, that's why it's working fine for Gadgetbridge

**Reply by pacjo** [reply]
@papjul do you have time to take another look?

**Reply by papjul** [reply]
> > Smartspacer handles registering the broadcast receiver for you, and will send your plugin's Broadcast Provider the raw intents so you can load any data from them, and act as required.
> 
> So intent should probably be sent to Smartspacer, and not GenericWeather.

If you want the intent to be sent to Smartspacer app, and Smartspacer app is taking care of registering receivers with intents, why isn't Smartspacer app showing in settings?

The relevant part of the code is really simple:
https://github.com/breezy-weather/breezy-weather/blob/v5.2.0/app/src/main/java/org/breezyweather/sources/RefreshHelper.kt#L820-L822

Maybe ask them? Maybe they have a bug where the intent is not correctly registered. I'm not an expert in this stuff, but the implementation of the Gadgetbridge intent works with Gadgetbridge apps.

Your first solution requires the additional permission to list all apps on device, and it would be more than confusing in the UI to list all apps, even the ones not supporting the intent, when only a few apps support it. And the second one, I'm not asking to add the Gadgetbridge intent manually on Smartspacer, but just to fix the way the intent is registered so that it shows up on our side, in a generic manner that would work with any intent.

Just to be clear: I'm not working on this as the feature is working as intended, that's why I moved it to discussions, but you can keep me informed of your progress or if you found that I need to make a small adjustment to the few lines of code I linked above.

**Comment by pacjo** [comment]
Ok, we've reached the conclusion that skipping Smartspacer functionality and implementing the broadcast receiver in the plugin itself is the way to go.

Not that I'm happy about it, but at least you'll have your privacy and there shouldn't be a need to change the code on your end.

I'll do further updates, when I get to it.

**Reply by papjul** [reply]
When you say "we", you mean you discussed it with Smartspacer developers?

**Reply by pacjo** [reply]
Yes

**Comment by pacjo** [comment, marked as answer]
Alright, I migrated to normal broadcast receiver and it's working fine since yesterday.

Commit with the change: [9588c2bbc3b9b5a48fa3952c4e7ace25cc72405a](https://github.com/pacjo/SmartspacerPlugins/commit/9588c2bbc3b9b5a48fa3952c4e7ace25cc72405a)

For anyone affected, fix will be available in the next release, but I don't know when that'll be. For now you can stick to last working version of breezy, compile from source or install (let's call it a pre-release) version attached (keep in mind there are other issues in this one).
[app-genericweather-release.zip](https://github.com/breezy-weather/breezy-weather/files/15015879/app-genericweather-release.zip)
